<launch>
  <!-- https://github.com/rickstaa/realsense-ros/blob/da38e309c2fbe88e45e6148bf5d944486b3e8e98/realsense2_description/launch/view_d435_model_rviz_gazebo.launch -->

  <!-- arguments for the urdf xacro file, rviz visualization and namespace -->
  <arg name="name" default="camera" />
  <arg name="topics_ns" default="camera" />
  <arg name="add_plug" default="false" />
  <arg name="publish_pointcloud" default="true" />
  <arg
    name="model"
    default="$(find visman_learning_adventure)/urdf/d435_camera.urdf.xacro"
  />

  <!-- Load the sdf model only the parameter server -->
  <param
    name="robot_description"
    command="$(find xacro)/xacro $(arg model)
      use_nominal_extrinsics:=true
      publish_pointcloud:=$(arg publish_pointcloud)
      add_plug:=$(arg add_plug)
      name:=$(arg name)
      topics_ns:=$(arg topics_ns)
      gravity:=0"
  />

  <node
    name="robot_state_publisher"
    pkg="robot_state_publisher"
    type="robot_state_publisher"
  >
    <param name="publish_frequency" type="double" value="5.0" />
  </node>

  <!-- https://github.com/pal-robotics/realsense_gazebo_plugin/issues/31#issuecomment-1098682872 -->
  <group ns="camera">
    <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" />

    <node
      pkg="nodelet"
      type="nodelet"
      name="nodelet1"
      args="load depth_image_proc/register nodelet_manager"
    >
      <remap from="rgb/camera_info" to="color/camera_info" />
      <remap from="depth/camera_info" to="depth/camera_info" />
      <remap from="depth/image_rect" to="depth/image_raw" />
      <remap
        from="depth_registered/camera_info"
        to="aligned_depth_to_color/camera_info"
      />
      <remap
        from="depth_registered/image_rect"
        to="aligned_depth_to_color/image_raw"
      />
    </node>
  </group>


  <!-- launch table + camera -->

  <param
    name="model_description"
    textfile="$(find visman_learning_adventure)/models/tabletop_aruco_test/description.sdf"
  />

  <!-- Run a python script to the send a service call to gazebo_ros to spawn a URDF robot -->
  <arg name="distance" default="0.3" />
  <node
    name="model_spawner"
    pkg="gazebo_ros"
    type="spawn_model"
    respawn="false"
    output="screen"
    args="-sdf -param model_description -model box"
  />

  <!-- now load the camera -->
    <node
    name="camera_spawner"
    pkg="gazebo_ros"
    type="spawn_model"
    respawn="false"
    output="screen"
    args="
        -urdf
        -param robot_description
        -model camera
        -unpause
        -x -2.0
        -y 0.0
        -z 1.0
        -P 0.2
    "
  />

  <!-- Load gazebo -->
  <arg name="paused" default="false" />
  <arg name="use_sim_time" default="true" />
  <arg name="headless" default="false" />
  <arg name="debug" default="false" />
  <arg name="gui" default="true" />

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
  <arg name="world_name" default="worlds/empty.world" />
    <arg name="debug" value="$(arg debug)" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="paused" value="$(arg paused)" />
    <arg name="use_sim_time" value="$(arg use_sim_time)" />
    <arg name="headless" value="$(arg headless)" />
  </include>

  <!-- Load rviz -->
  <node
    name="rviz"
    pkg="rviz"
    type="rviz"
    args="-d $(find realsense2_description)/rviz/urdf.rviz"
    required="true"
    if="$(arg gui)"
  />

</launch>
